components/script/dom/htmlinputelement.rs

```diff
@@ -59,6 +59,66 @@ enum InputType {
     InputPassword
 }
 
+#[derive(Debug, PartialEq)]
+enum InputValueMode {
+    Value,
+    Default,
+    DefaultOn,
+    Filename,
+}
+
+impl Default for InputValueMode {
+    fn default() -> InputValueMode { InputValueMode::Default }
+}
+
+impl InputValueMode {
+
+    // https://html.spec.whatwg.org/multipage/#input-type-attr-summary
+    fn from_input_element(input: &HTMLInputElement) -> InputValueMode {
+        match input.input_type.get() {
+            InputType::InputSubmit
+                | InputType::InputReset
+                | InputType::InputButton
+                | InputType::InputImage => Default::default(),
+            InputType::InputCheckbox
+                | InputType::InputRadio => InputValueMode::DefaultOn,
+            InputType::InputPassword
+                | InputType::InputText => InputValueMode::Value,
+            InputType::InputFile => InputValueMode::Filename,
+        }
+    }
+
+    fn get_value(&self, input: &HTMLInputElement) -> DOMString {
+        let default = match *self {
+            InputValueMode::DefaultOn => DOMString::from("on"),
+            InputValueMode::Filename => DOMString::from(r"C:\fakepath\"),
+            _ => DOMString::from(""),
+        };
+
+        if *self == InputValueMode::Value {
+            input.textinput.borrow().get_content()
+        } else {
+            input.upcast::<Element>()
+                .get_attribute(&ns!(), &atom!("value"))
+                .map_or_else(|| default,
+                             |a| DOMString::from(a.summarize().value))
+        }
+    }
+
+    fn set_value(&self, input: &HTMLInputElement, value: DOMString) {
+        let element = input.upcast::<Element>();
```

Comment:
I see no value in putting the let-binding here, since `element` is only used in the else-body.